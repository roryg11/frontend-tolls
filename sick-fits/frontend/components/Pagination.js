    import React, { Component } from 'react';
    import PaginationStyles from "./styles/PaginationStyles";
    import { Query } from "react-apollo";
    import gql from "graphql-tag";
    import { perPage } from "../config.js" 
    import Link from "next/link";
    import Head from "next/head";

    const PAGINATION_QUERY = gql`
        query PAGINATION_QUERY {
            itemsConnection {
                aggregate {
                    count
                }
            }
        }
    `
    
    class Pagination extends Component {
        render() {
            return (
                    <Query query={ PAGINATION_QUERY }>
                        { ({data, loading, error})=>{
                            if(loading) return <p>Loading...</p>; 
                            const count = data.itemsConnection.aggregate.count;
                            const pages = Math.ceil(count/perPage);
                            return (<PaginationStyles>
                                <Head>
                                    <title>Sick Fits | {this.props.page} of {pages}</title>
                                </Head>
                                <Link prefetch
                                href={{
                                        pathName: "/items",
                                        query: {
                                            page: this.props.page -1
                                        }
                                    }}><a>Prev</a></Link>
                                <p>{this.props.page} of {pages}</p>
                                <Link prefetch
                                href={{
                                        pathName: "/items",
                                        query: {
                                            page: this.props.page + 1
                                        }
                                    }}><a>Next</a></Link>
                            </PaginationStyles>);

                        } }
                    </Query>
            );
        }
    }
    
    export default Pagination;