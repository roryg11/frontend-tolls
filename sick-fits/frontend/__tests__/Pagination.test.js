import {shallow, mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { MockedProvider } from "react-apollo/test-utils";
import wait from "waait";
import Router from "next/router";
import Pagination, { PAGINATION_QUERY } from "../components/Pagination";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

Router.router = {
    push(){},
    prefetch(){}
}

function makeMocksFor(length){
    return [{
        request: {query: PAGINATION_QUERY},
        result: { 
            data: { 
                itemsConnection: {
                    __typename: 'aggregate',
                    aggregate: {
                        __typename: 'count',
                        count: length
                    }
                }
            }
        }
    }];
}

describe('<Pagination/>', ()=>{
    it('displays a loading message', ()=>{
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        expect(wrapper.text()).toContain("Loading...");
    });
    it('loads and renders', async ()=>{
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const total = wrapper.find('.totalPages');
        expect(total.text()).toEqual("5");
        const pagination = wrapper.find('div[data-test="pagination"]');
        expect(toJSON(pagination)).toMatchSnapshot();
    });

    it('disables prev link for first page', async ()=>{
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop("aria-disabled")).toEqual(true);
        expect(wrapper.find('a.next').prop("aria-disabled")).toEqual(false);
    });

    it('disables the next link for the last page', async ()=>{
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={5}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop("aria-disabled")).toEqual(false);
        expect(wrapper.find('a.next').prop("aria-disabled")).toEqual(true);
    });

    it('enables both buttons on middle pages', async ()=>{
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={3}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop("aria-disabled")).toEqual(false);
        expect(wrapper.find('a.next').prop("aria-disabled")).toEqual(false);
    });
})