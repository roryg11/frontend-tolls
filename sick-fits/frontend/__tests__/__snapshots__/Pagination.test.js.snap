// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> loads and renders 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathName": "/items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="?page=0"
      onClick={[Function]}
    >
      Prev
    </a>
  </Link>
  <p>
    1
     of 
    <span
      className="totalPages"
    >
      5
    </span>
  </p>
  <Link
    href={
      Object {
        "pathName": "/items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      className="next"
      href="?page=2"
      onClick={[Function]}
    >
      Next
    </a>
  </Link>
</div>
`;
